### DISCUSSION TITLE
LexicalNegation, but for all operators

### DISCUSSION THREAD

**Discussion by JakobBruenker** [ghc-proposals/ghc-proposals#444] [Ideas]
GHC has the `-XPostfixOperators` extension, which makes it possible to write thing like
```haskell
ghci> (!) n = product [1..n]
ghci> (5!)
120
```
However, despite being around for a while, this extension is still used relatively rarely.

I believe this is because having to use parentheses puts people off, and `(5!)` isn't much better to type or look at than `fac 5`.

More recently, `-XLexicalNegation` was introduced, which means `-5` and `(-5)` are interpreted as negative 5, but `(- 5)` is interpreted as `\x -> x - 5`. At around the same time, whitespace-sensitivity was introduced or made more consistent for other symbols, such as `@` and `!`.

There could be an opportunity here: It seems possible interpret `x + y` as before, but interpret both `x+` and `+x` as `(+) x`. If the precedences work as with `-XLexicalNegation`, this would give us prefix- and postfix-operators that never need parentheses around them.

(As a couple of further clarifying examples, `x +y` would be interpreted as `x ((+) y)`, and `x+ y` as `((+) x) y` - which happens to be the same as `x + y`. `x+y` presumably would also be interpreted the same way as `x + y` - one could imagine giving this higher precedence than `x + y`, but that seems like it might be too much of a change.)

(Thanks to Richard for this suggestion:)
We could introduce some fairly lightweight syntax to allow different definitions for the different possible locations a symbol can have, like
```haskell
(+) = ... -- regular infix symbol
(_+) = ... -- postfix symbol
(+_) = ... -- prefix symbol
```

This proposal would introduce more whitespace-sensitivity, which can be controversial, however, it would fit in well with the other recently introduced changes doing the same.

Nonetheless, it is a syntax extension that doesn't make the language fundamentally more expressive, and thus would undoubtedly be a controversial extension. Furthermore, most useful examples that come to my mind are from various mathematical domains, and would ideally be using Unicode symbols as well, which is controversial in itself. Because of this, I wanted to gauge interest/opposition here before deciding whether pursuing a proposal would be worth it.

**Comment by nomeata** [comment]
I have a warm spot in my heart for more whitespace sensitivity to get rid of parentheses, and if I ever get to design a language I'll go wild on them. I particular the higher preference of `x+y` is something that I'd probably love.

But for Haskell as it stands right now, I am hesitant. Operator sections without parentheses is a bit “quiet” for something as “big” as turning something into a function. So I'm all for exploring and discussing, but would warn again actually making such changes without stronger justification.

**Comment by JakobBruenker** [comment]
I just realized the title may be slightly misleading by the way, since one thing that `LexicalNegation` does that's not possible with the top comment as it stands is that the same operator can have different meanings when used as prefix vs as infix operator

**Reply by int-index** [reply]
I can imagine writing `a + b` and `a * b` for numeric operations, and at the same time `p+` and `p*` instead of `some p` and `many p` in parser combinators (to mimic [EBNF](https://www.w3.org/TR/REC-xml/#sec-notation)).

But if the infix/postfix usage should use the same definition, I can’t actually think of a compelling use case.

**Reply by JakobBruenker** [reply]
That's fair; It would definitely be preferable to allow three separate definitions for the three different usages, but I don't have a good idea at the moment for how to actually specify separate definitions. (i.e. the ideas I do have seem too ad-hoc and unlikely to be popular.)

Edit: Though I think what Richard suggested below would be a good approach, i.e. having `(+)`, `(_+)`, and `(+_)`

**Comment by Anrock** [comment]
I have a feeling that it will really confuse beginners. Honestly I guess it will confuse me too - one whitespace may change the meaning of expression especially for non-commutative operators.

While it's a way more drastic change than subj, I believe Haskell should adapt mixfix syntax for defining operators like Agda/Idris.

**Comment by Ericson2314** [comment]
Somewhat relatedly, I wish we could opt into Agda's (IIUC) simplified lexing where everything needs to be space separated except for leading and trailing tokens.

x+y would lex as a single token, and to (optionally) go in a different direction than this, we could use that to write a better error message that this is an invalid identifier (since it is not valid infix op or regular alphanumeric identifier). So a more conservative language but with better error messages. Maybe could be controlled with `-XFlexibleAdjecentTokens` and `-XNoFlexibleAdjecentTokens`, with former being today's behavior and the default.

**Comment by goldfirere** [comment]
One minor point on the original idea: it describes `+x` as meaning `(+) x`. This means that `(+x)` and `(+ x)` have different meanings! (The latter means `\y -> (+) y x`.)

I'm somewhat sympathetic to the desire to allow different definitions for different occurrence locations of a symbol, but I think this is a step too far, myself. But if we did do it, we could have

```
(!) :: Vector a -> Int -> Maybe a  -- lookup
(_!) :: Int -> Int -- factorial
```

I've taken advantage of the fact that `_` cannot be used in symbols.

**Reply by JakobBruenker** [reply]
Yes, using underscores like this would like a good approach, if we did do it - somewhat reminiscent of Agda.

`(+x)` and `(+ x)` being different is actually exactly the situation we have with `-XLexicalNegation` now, with `(-x)` and `(- x)`.

**Comment by int-index** [comment]
If we extended this to `[]`, then we could have `xs[i]` instead of `xs !! i`. Seems quite fitting now that we have record dot syntax.
